File: algorithms/implementation/between-two-sets/between-two-sets-javascript/js/getTotalX.js

// Função que recebe dois arrays e retorna a quantidade de números inteiros entre os valores máximos e mínimos dos arrays,
// que são múltiplos de todos os elementos do primeiro array e divisores de todos os elementos do segundo array.
// export function getTotalX(arr, brr) {
    
//     // Inicializa o resultado como 0.
//     let resultado = 0;

//     // Loop que percorre todos os números inteiros no intervalo entre o máximo de arr e o mínimo de brr.
//     for (let index = Math.max(...arr); index <= Math.min(...brr); index++) {

//         // Inicializa a variável como verdadeira.
//         let ehFatorMultiplo = true;

//         // Loop que verifica se index é múltiplo de todos os elementos de arr.
//         for (const elemento of arr) {
            
//             // Se não for múltiplo, define ehFatorMultiplo como falso e interrompe o loop.
//             if (index % elemento != 0) {
                
//                 ehFatorMultiplo = false;
//                 break;
//             }
//         }
        
//         // Se index for múltiplo de todos os elementos de arr, continua a verificação com brr.
//         if (ehFatorMultiplo) {
            
//             // Loop que verifica se index é divisor de todos os elementos de brr.
//             for (const elemento of brr) {
                
//                 // Se não for divisor, define ehFatorMultiplo como falso e interrompe o loop.
//                 if (elemento % index != 0) {
                    
//                     ehFatorMultiplo = false;
//                     break;
//                 }
//             }
//         }

//         // Se index for múltiplo de todos os elementos de arr 
//         // e divisor de todos os elementos de brr, incrementa o resultado.
//         if (ehFatorMultiplo) {
            
//             resultado++;
//         }
//     }

//     // Retorna o resultado final.
//     return resultado;
// }

/**
 * Retorna o número total de inteiros que são múltiplos de todos os elementos de 'arr' e
 * fatores de todos os elementos de 'brr' no intervalo de 1 até o primeiro elemento de 'brr'.
 *
 * @param {Array} arr - Lista de inteiros para verificar os múltiplos.
 * @param {Array} brr - Lista de inteiros para verificar os fatores.
 * @return {number} O número total de inteiros que atendem aos critérios.
 */
export function getTotalX(arr, brr) {

    // Inicializa a variável para armazenar o número total de inteiros que atendem aos critérios.
    var partidas = 0;

    // Loop que percorre todos os inteiros de 1 até o primeiro elemento de 'brr'.
    for (let index = 1; index <= brr[0]; index++) {
        
        // Verifica se index é múltiplo de todos os elementos de 'arr' usando o método reduce.
        const multiplo_elementoA = arr.reduce((m, e) => m && index % e == 0, true);

        // Verifica se index é fator de todos os elementos de 'brr' usando o método reduce.
        const fator_elementoB = brr.reduce((m, e) => m && e % index == 0, true);

        // Incrementa a contagem se index for múltiplo de 'arr' e fator de 'brr'.
        partidas += multiplo_elementoA && fator_elementoB ? 1 : 0;

    }

    // Retorna o número total de inteiros que atendem aos critérios.
    return partidas
}

File: algorithms/implementation/between-two-sets/between-two-sets-python/tests/getTotalX.py
# No problema "Between Two Sets" 
# do HackerRank, você é solicitado 
# a encontrar e contar todos os números 
# inteiros que são divisíveis por todos 
# os números de um conjunto e que, por 
# sua vez, dividem uniformemente todos 
# os números de outro conjunto. Em outras 
# palavras, você precisa encontrar os 
# inteiros que são fatores comuns a 
# ambos os conjuntos.

# def getTotalX(arr, brr):
    
#     # Encontrar o maior elemento em 'arr' e o menor em 'brr'
#     max_arr = max(arr)
#     min_brr = min(brr)
    
#     # Inicializar o contador para a contagem de números que atendem às condições
#     contador = 0
    
#     # Iterar pelos números potenciais candidatos
#     for i in range(max_arr, min_brr + 1):
        
#         # Verificar se 'i' é divisível por todos os elementos de 'arr'
#         if all(i % item == 0 for item in arr):
            
#             # Verificar se todos os elementos de 'brr' são divisíveis por 'i'
#             if all(item % i == 0 for item in brr):
                
#                 # Incrementar o contador se ambas as condições forem atendidas
#                 contador += 1
                
#     # Retornar o contador final
#     return contador


"""
Retorna o número total de inteiros que são múltiplos de todos os elementos de 'arr' e
fatores de todos os elementos de 'brr' no intervalo definido pelas operações gcd e lcm.
"""
# Importa o módulo de matemática para utilizar as funções gcd e lcm.

import math

def getTotalX(arr, brr):
    
    # Calcula o Maior Comum Divisor (gcd) de 'brr'.
    minBrr = math.gcd(*brr)
    
    # Calcula o Mínimo Múltiplo Comum (lcm) de 'arr'.
    maxArr = math.lcm(*arr)
    
    # Inicializa a variável para armazenar o número total de inteiros que atendem aos critérios.
    total = 0
    
    # Loop que continua enquanto o Mínimo Comum Divisor (minBrr) é maior ou igual ao Mínimo Múltiplo Comum (maxArr).
    while minBrr >= maxArr:
        
        # Inicializa uma lista para armazenar os resultados.
        resultado = []
        
        # Loop que verifica se minBrr é múltiplo de maxArr e fator de todos os elementos de 'brr'.
        for i in brr:
            
            if(minBrr % maxArr == 0) and (i % minBrr == 0):
                
                # Adiciona minBrr à lista de resultados.
                resultado.append({ minBrr })
                
        # Verifica se o tamanho da lista de resultados é igual ao tamanho de 'brr'.
        if(len(resultado) == len(brr)):
            
            # Incrementa o total se todos os elementos de 'brr' atendem aos critérios.
            total += 1
            
        # Decrementa minBrr pelo valor de maxArr.
        minBrr -= maxArr
    
    # Retorna o número total de inteiros que atendem aos critérios.
    return total